recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb

```diff
@@ -0,0 +1,505 @@
+        "  \n",
+        "With all the preparations complete, we now start the training process. The model will be exposed to numerous examples from our legal dataset, gradually adjusting its understanding of legal concepts.  \n",
+        "  \n",
+        "We'll monitor the training loss over time, which should decrease as the model improves its performance on the task. After training, we'll save the fine-tuned model for future use.  "
+      ],
+      "metadata": {
+        "id": "uKIe2z_THh35"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "start_time = timeit.default_timer()\n",
+        "# Start training\n",
+        "trainer.train()\n",
+        "training_time = timeit.default_timer() - start_time\n"
+      ],
+      "metadata": {
+        "id": "oZ2rUKO2sdae"
+      },
+      "execution_count": null,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "## Saving the Fine-tuned Model  \n",
+        "  \n",
+        "After the training process is complete, it's crucial to save our fine-tuned model. This step ensures that we can reuse the model later without having to retrain it. We'll save both the model weights and the tokenizer, as they work in tandem to process and generate text.  \n",
+        "  \n",
+        "Saving the model allows us to distribute it, use it in different environments, or continue fine-tuning it in the future. It's a critical step in the machine learning workflow, preserving the knowledge our model has acquired through the training process.  "
+      ],
+      "metadata": {
+        "id": "zUjBCOZVIaas"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "model.save_pretrained(\"./results\")\n",
+        "tokenizer.save_pretrained(\"./results\")"
+      ],
+      "metadata": {
+        "id": "rK1i3CFeebzu"
+      },
+      "execution_count": null,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "### Persisting the Model to Huggingface Hub  \n",
+        "  \n",
+        "After fine-tuning and validating our model, a optional step is to make it easily accessible for future use or sharing with the community. The Hugging Face Hub provides an excellent platform for this purpose.  \n",
+        "  \n",
+        "Uploading our model to the Hugging Face Hub offers several benefits:  \n",
+        "1. Easy sharing and collaboration with other researchers or developers  \n",
+        "2. Version control for your model iterations  \n",
+        "3. Integration with various libraries and tools in the Hugging Face ecosystem  \n",
+        "4. Simplified deployment options  \n",
+        "  \n",
+        "We'll demonstrate how to push our fine-tuned model and tokenizer to the Hugging Face Hub, making it available for others to use or for easy integration into other projects. This step is essential for reproducibility and for contributing to the broader NLP community.  "
```

Comment:
Can we talk about the licensing considerations here? What are the publishing requirements given the data license and the model license?

Reply:
The dataset is ASFv2 and we're not publishing, so I'm 99.999% sure it's safe.

Reply:
Ok, should we remove the "making it available for others to use" language? It seems we should at least alert users to copyright concerns otherwise.

Reply:
Yea, I can drop the 'making it available for others to use' language, nbd. But can you expand on copyright concerns? I'm asking in good faith, I truly don't see any- it's an ASFv2 licensed model fine tuned on ASFv2 licensed data.

Reply:
Sorry, I mean bring up copyright concerns for when someone wants to apply this pattern to a different dataset. You could say that that "When considering publishing a fine-tuned model, models and datasets should be evaluated for copyright constraints. Here we are using an ASFv2 licensed model fine-tuned on ASFv2 licensed data." or something similar.